cranelift/codegen/src/egraph.rs

```diff
@@ -166,6 +169,11 @@ where
             if let NewOrExistingInst::Existing(inst) = inst {
                 debug_assert_eq!(self.func.dfg.inst_results(inst).len(), 1);
                 let result = self.func.dfg.first_result(inst);
+                debug_assert_eq!(
+                    self.get_available_block(inst),
+                    self.available_block[orig_result]
+                );
```

Comment:
This debug-assert is failing in a few of the wasm spec tests and I don't understand why yet.

In every case that's failed, the GVN map says the instruction we looked up is equivalent to an existing instruction, but the available-block for the existing instruction is the entry block, while the available-block for the new instruction is somewhere deeper in the dominator tree.

I believed that GVN would always yield the same available-block for pure instructions so this is concerning.

Reply:
I'm still reasoning through this. For GVN to match, both instructions must be identical, except that each operand is only required to be in the same set in the union-find rather than actually equal. The assertion failing means that for some operand in the instruction, the two instructions consume values which are in the same union-find set but have different available-blocks.

I had been concerned about this specific situation but Chris and I reasoned that it was okay (https://github.com/bytecodealliance/wasmtime/pull/7891#issuecomment-1960531054) and now I'm confused about whether we were right and the assertion is overly strict. If so, what is the correct available-block to use here?

Reply:
I suspect what is happening here is that the instruction gets rewritten the first time it is seen and the GVN-map entry points to the rewritten result (which could be e.g. an `iconst` available everywhere), even though the instruction per its arguments is only available in one block. So I think the assertion that we want is that the available block of `orig_result` (which we're going to use as our value here) dominates the available block of `inst`, rather than is exactly the same. I think that resolves the issue?

Reply:
That does make the assert pass in all the tests we have. I think you're correct but the reason why was obscured by the surrounding code.

First off, the GVN-hit case did not always return `orig_result`. If we were looking up a `NewOrExistingInst::Existing` instruction, `insert_pure_enode` actually returned the result of that instruction. But that case only happens when this function was called from `remove_pure_and_optimize`, which ignores the return value. The return value only matters for `make_inst_ctor`, which always passes `NewOrExistingInst::New`, in which case we were returning `orig_result`.

Therefore it's safe to change this function to always return `orig_result`, which I think is what we probably intended in the first place, and based on your comment is what you thought we were already doing. :grin:

The next observation is that if the result of `inst` never appears in the values of the `value_to_opt_value` map, then it will never get looked up in `available_block` (or in the union-find, for that matter). So we don't actually need an answer to the question of what block this instruction should be available in. In fact it's better to leave it at the "reserved" default value so we'll get assertion failures if we accidentally do look it up.

I've made these changes; could you review again when you have a chance? Hopefully CI will pass now too.